packages/graph/hash_graph/lib/graph/src/types/schema/array.rs

```diff
@@ -0,0 +1,313 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum ArrayTypeTag {
+    Array,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+struct Array {
+    r#type: ArrayTypeTag,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    min_items: Option<usize>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    max_items: Option<usize>,
+}
+
+impl Array {
+    #[must_use]
+    pub fn new(min_items: impl Into<Option<usize>>, max_items: impl Into<Option<usize>>) -> Self {
+        Self {
+            r#type: ArrayTypeTag::Array,
+            min_items: min_items.into(),
+            max_items: max_items.into(),
+        }
+    }
+
+    #[must_use]
+    pub const fn min_items(&self) -> Option<usize> {
+        self.min_items
+    }
+
+    #[must_use]
+    pub const fn max_items(&self) -> Option<usize> {
+        self.max_items
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct TypedArray<T> {
+    items: T,
+    #[serde(flatten)]
+    array: Array,
+}
+
+impl<T> TypedArray<T> {
+    #[must_use]
+    pub fn new(
+        items: T,
+        min_items: impl Into<Option<usize>>,
+        max_items: impl Into<Option<usize>>,
+    ) -> Self {
+        Self {
+            items,
+            array: Array::new(min_items, max_items),
+        }
+    }
+
+    #[must_use]
+    pub const fn items(&self) -> &T {
+        &self.items
+    }
+
+    #[must_use]
+    pub const fn min_items(&self) -> Option<usize> {
+        self.array.min_items()
+    }
+
+    #[must_use]
+    pub const fn max_items(&self) -> Option<usize> {
+        self.array.max_items()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct MaybeOrderedArray {
```

Comment:
I don't have a suggestion yet but I'm not sure about the naming. Feels strange to me that we don't encapsulate the "Maybe" stuff about min/max but we do about Ordering. Also when reading it I was reading it like "Maybe OrderedArray".

Reply:
How do you feel about your latter solution? It could let us end up in some invalid states, like I'm not sure we should be allowing untyped arrays in the first place (outside of the `Empty List` data type) so `MaybeOrdered<Array>` doesn't make sense to me. I think we could stop it being composed like that though.

Could we perhaps have `Array<Undefined>` as a special case rather than needing `Itemized<Array, T>` for the others?

Reply:
I'm fine with `MaybeOrdered<Itemized<Array, T>>` though. Lets swap to that style